ffi/src/Aten/Cast.hs

```diff
@@ -0,0 +1,430 @@
+
+instance Castable CULLong CULLong where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable CULong CULong where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable CUSeconds CUSeconds where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable CUShort CUShort where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable CWchar CWchar where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Int8 Int8 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Int16 Int16 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Int32 Int32 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Int64 Int64 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Word8 Word8 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Word16 Word16 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Word32 Word32 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable Word64 Word64 where
+  cast x f = f x
+  uncast x f = f x
+
+instance Castable (Ptr CInt) (Ptr CInt) where 
+  cast x f = f x
+  uncast x f = f x 
+
+instance Castable (Ptr CChar) (Ptr CChar) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr CUInt) (Ptr CUInt) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr CULong) (Ptr CULong) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr CLong) (Ptr CLong) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr CDouble) (Ptr CDouble) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr CString) (Ptr CString) where 
+  cast x f = f x 
+  uncast x f = f x 
+
+instance Castable (Ptr ()) (Ptr ()) where
+  cast x f = f x
+  uncast x f = f x
+
+
+instance Castable Int CInt where
+  cast x f = f (fromIntegral x)
+  uncast x f = f (fromIntegral x)
+
+instance Castable Int16 CShort where
+  cast x f = f (fromIntegral x)
+  uncast x f = f (fromIntegral x)
+
+instance Castable Int8 CChar where
+  cast x f = f (fromIntegral x)
+  uncast x f = f (fromIntegral x)
+
+instance Castable Word CUInt where
+  cast x f = f (fromIntegral x)
+  uncast x f = f (fromIntegral x)
+
+instance Castable Word8 CChar where
+  cast x f = f (fromIntegral x)
+  uncast x f = f (fromIntegral x)
+  
+instance Castable Double CDouble where
+  cast x f = f (realToFrac x)
+  uncast x f = f (realToFrac x)
+
+instance Castable [Double] (Ptr CDouble) where
+  cast xs f = newArray (map realToFrac xs) >>= f 
+  uncast xs f = undefined 
+
+instance Castable [Int] (Ptr CInt) where
+  cast xs f = newArray (map fromIntegral xs) >>= f
+  uncast xs f = undefined 
+
+instance Castable ByteString CString where
+  cast x f = useAsCString x f
+  uncast x f = packCString x >>= f 
+
+
+instance Castable [ByteString] (Ptr CString) where
+  cast xs f = do ys <- mapM (\x -> useAsCString x return) xs
+                 withArray ys $ \cptr -> f cptr
+  uncast xs f = undefined 
+
+class CppObject a where
```

Comment:
`class CppObject` and `instance (CppObject a) => Castable (ForeignPtr a) (Ptr a)` are casts for managed codes.